packages/mdc-select/index.ts

```diff
@@ -0,0 +1,596 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+import {CustomEventListener, MDCComponent} from '@material/base/index';
+import {EventType, SpecificEventListener} from '@material/base/index';
+import {MDCFloatingLabel} from '@material/floating-label/index';
+import {MDCLineRipple} from '@material/line-ripple/index';
+import * as menuSurfaceConstants from '@material/menu-surface/constants';
+import * as menuConstants from '@material/menu/constants';
+import {MDCMenu, MenuItemEvent} from '@material/menu/index';
+import {MDCNotchedOutline} from '@material/notched-outline/index';
+import {MDCRipple, MDCRippleFoundation, RippleCapableSurface} from '@material/ripple/index';
+import {cssClasses, strings} from './constants';
+import {MDCSelectFoundation} from './foundation';
+import {MDCSelectHelperText} from './helper-text';
+import {MDCSelectIcon} from './icon';
+import {
+  FoundationMapType,
+  HelperTextFactory, IconFactory,
+  LabelFactory,
+  LineRippleFactory, MenuFactory,
+  OutlineFactory, SelectEventDetail,
+} from './types';
+
+const VALIDATION_ATTR_WHITELIST = ['required', 'aria-required'];
+const MOUSEDOWN_TOUCHSTART_EVENTS: EventType[] = ['mousedown', 'touchstart'];
+
+class MDCSelect extends MDCComponent<MDCSelectFoundation> implements RippleCapableSurface {
+  static attachTo(root: Element): MDCSelect {
+    return new MDCSelect(root);
+  }
+
+  // Public visibility for this property is required by RippleCapableSurface.
+  root_!: HTMLElement; // assigned in MDCComponent constructor
+
+  private menu_!: MDCMenu | null; // assigned in enhancedSelectSetup_()
+  private menuOpened_ = false;
+
+  // Exactly one of these fields must be non-null.
+  private nativeControl_!: HTMLSelectElement | null;
+  private selectedText_!: HTMLElement | null;
+
+  private hiddenInput_!: HTMLInputElement | null;
+  private leadingIcon_!: MDCSelectIcon | null;
+  private helperText_!: MDCSelectHelperText | null;
+  private menuElement_!: Element | null;
+  private ripple!: MDCRipple | null;
+  private lineRipple_!: MDCLineRipple | null;
+  private label_!: MDCFloatingLabel | null;
```

Comment:
Similar comment here to one I left on #4377 - do all of these need to be `!` and `| null` or can some or all of them simply be left as undefined if the element isn't present?

Reply:
They can be left undefined, but the `if` statements are a little more verbose than a ternary. It's really just a style choice.